client/checker_client/bigmap.py

```diff
@@ -0,0 +1,162 @@
+        if diff["action"] != "update":
+            continue
+        if bigmap_id not in keys:
+            keys[bigmap_id] = set([])
+        for update in diff["updates"]:
+            key = micheline_value_to_python_object(update["key"])
+            # If the update removes the key value pair, remove it from our map
+            # TODO: I am pretty sure this is safe for cases where the value gets updated to
+            # an Option with value None, but am not 100% sure.
+            if micheline_value_to_python_object(update["value"]) is None:
+                keys[bigmap_id].remove(key)
+            # Otherwise record the key
+            else:
+                keys[bigmap_id].add(key)
+    return keys
+
+
+def _render_bigmap(
+    field: str, bigmap_id: int, keys: Dict[int, Set[Any]], contract: ContractInterface
+):
+    full_map = {}
+    bigmap_id = str(bigmap_id)
+    if bigmap_id not in keys:
+        return {}
+    for key in keys[bigmap_id]:
+        # Call tezos to get the value for this key
+        full_map[key] = contract.storage[field][key]()
+    return full_map
+
+
+def storage_with_bigmaps(
+    contract: ContractInterface,
+    contract_bigmap_keys: Dict[int, Set[Any]],
+) -> Any:
+    """Queries a contract's storage, rendering bigmaps with their full contents.
+
+    Warning: since this method forces big_maps which are usually lazy, it will load all of
+    their contents into memory.
+
+    Args:
+        contract: The contract of interest
+        contract_bigmap_keys: The keys in this contract's bigmaps. Missing bigmap_ids will be loaded as
+            empty dicts.
+    Returns:
+        A python representation of the contract's storage (similar to calling contract.storage())
+    """
+
+    def _rec_load_storage(contract, contract_bigmap_keys, michelson_data):
+        storage = {}
+        # This node is a big map, let's populate it with its data
+        if isinstance(michelson_data, BigMapType):
+            return _render_bigmap(
+                michelson_data.field_name,
+                michelson_data.to_python_object(),
+                contract_bigmap_keys,
+                contract,
+            )
+        # This node is not a big map and has no children, just return it
+        # Currently this also catches lists, which means that bigmaps nested within a list
+        # will not be rendered.
+        if not hasattr(michelson_data, "items") or isinstance(michelson_data, ListType):
+            # Only render items which are not "undefined". Pytezos uses same behavior internally.
+            if michelson_data is Undefined:
+                return Undefined
+            return michelson_data.to_python_object()
+        # Otherwise, this element is a container with child elements, recurse into them
+        if isinstance(michelson_data, ADTMixin):
+            # Because reasons, need to select the second element from this generator
+            child_iterator = map((lambda e: e[1]), michelson_data.iter_values())
```

Comment:
Hahaha, nice :stuck_out_tongue: If there is any source for the "reasons", it might be a good idea to drop a link/explanation here.

Reply:
Haha very true. In this case it is purely an API difference between the two iterators, so I'm not sure if it is worth documenting further.